modules/packages/ConcurrentMap.chpl

```diff
@@ -0,0 +1,1046 @@
+      var tok = getToken();
+      tok.pin();
+      // Data stored on recursionStack: PointerList, start, i
+      var recursionStack = new LockFreeStack(stackType);
+      var deferred : unmanaged DeferredNode(deferredType)?;
+      var restore = true;
+      var curr : unmanaged Buckets(keyType, valType)? = root;
+      var start = ((iterRNG.getNext())%(curr!.buckets.size):uint):int;
+      var startIndex = 0;
+
+      while (true) {
+        restore = true;
+        for i in startIndex..(curr!.buckets.size-1) {
+          var idx = (start + i)%curr!.buckets.size;
+          var bucketBase = curr!.buckets[idx].read();
+          if (bucketBase != nil) {
+            if (bucketBase!.lock.read() == E_AVAIL && bucketBase!.lock.compareAndSwap(E_AVAIL, E_LOCK)) {
+              var bucket = bucketBase : unmanaged Bucket(keyType, valType)?;
+              for j in 0..#bucket!.count do yield (bucket!.keys[j], bucket!.values[j]);
+              bucket!.lock.write(E_AVAIL);
+            } else if (bucketBase!.lock.read() == P_INNER) {
+              var stackElem = (curr, start, i);
+              recursionStack.push(stackElem);
+              curr = bucketBase : unmanaged Buckets(keyType, valType)?;
+              start = ((iterRNG.getNext())%(curr!.buckets.size):uint):int;
+              startIndex = 0;
+              restore = false;
+              break;
+            } else {
+              var deferredElem = (curr, idx);
+              var deferredNode = new unmanaged DeferredNode(deferredElem);
+              deferredNode.next = deferred;
+              deferred!.prev = deferredNode;
+              deferred = deferredNode;
+            }
+          }
+        }
+
+        if (restore == true) {
+        var (hasState, state) = recursionStack.pop();
+          if (hasState) {
+            curr = state[0];
+            start = state[1];
+            startIndex = state[2] + 1;
+            continue;
+          } else {
+            var head = deferred;
+            var continueFlag = false;
+            while (head != nil) {
+              var pList = head!.val[0];
+              var idx = head!.val[1];
+              var bucketBase = pList!.buckets[idx].read();
+              var next = head!.next;
+              if (bucketBase!.lock.read() == P_INNER) {
+                delete head;
+                curr = bucketBase : unmanaged Buckets(keyType, valType)?;
+                start = ((iterRNG.getNext())%(curr!.size):uint):int;
+                startIndex = 0;
+                continueFlag = true;
+                break;
+              } else if (bucketBase!.lock.read() == E_AVAIL && bucketBase!.lock.compareAndSwap(E_AVAIL, E_LOCK)) {
+                delete head;
+                var bucket = bucketBase : unmanaged Bucket(keyType, valType)?;
+                for j in 0..#bucket!.count do yield (bucket!.keys[j], bucket!.values[j]);
+                bucket!.lock.write(E_AVAIL);
+              }
+              head = next;
```

Comment:
After I adjusted line `430` I started spinning forever. You know this logic better than I do, but it seems to me we'd want to update `deferred` after consuming it (else we'll keep looping forever down below).

After I tried doing that I got stuck spinning on a CAS.